### PR TITLE
Update cluster management doc.

### PR DESCRIPTION
**Pull request by gmarek**
Can someone take a look if this is what was needed for #10125? I was lacking a bit of context.

### CODE REVIEW

**Comment by wojtek-t** [comment]
Ref #10125

**Comment by wojtek-t** [comment]
cc @thockin

**Comment by gmarek** [comment]
@thockin @wojtek-t @jszczepkowski @alex-mohr

**Comment by gmarek** [comment]
@wojtek-t You could have given me few minutes to flesh up this PRs description:)

File: docs/cluster_management.md [hunk 1]
```diff
@@ -2,45 +2,76 @@
...
 
-Specific API versions can be turned on or off by passing --runtime-config=api/<version> flag while bringing up the server. For example: to turn off v1 API, pass --runtime-config=api/v1=false.
-runtime-config also supports 2 special keys: api/all and api/legacy to control all and legacy APIs respectively. For example, for turning off all api versions except v1, pass --runtime-config=api/all=false,api/v1=true.
+Two interesting examples:
+* `max-pods` in Kubelet setting maximal number of Pods that are allowed to run on a single Node,
+* `pod-eviction-timeout` in kube-controller-manager setting how long we wait after Node becomes unresponsive before we start evicting Pods from it.
+More can be seen by running particulars app help.
 
-### Switching your cluster's storage API version
+If you want to change the value of particular flag, e.g. `max-pod`, you need to find a file used by Salt to set Kubelet flags. This particular file is `cluster/saltbase/salt/kubelet/default`. In this file you can modify the line with DAEMON_ARGS by adding the value for the flag you want, e.g. change the DAEMON_ARGS line to
```
Comment (thockin) [inline docs/cluster_management.md]: We should be careful not to spec too much about salt.  Only some of the platforms use salt.  Also, we should not replicate too much of the cluster turnup docs into here.  Did you take a look at Eric's new 'from scratch' guide?  Can we remove anything here in favor of putting it there?

File: docs/cluster_management.md [hunk 2]
```diff
@@ -2,45 +2,76 @@
...
+```
 
-KUBE_API_VERSIONS env var controls the API versions that are supported in the cluster. The first version in the list is used as the cluster's storage version. Hence, to set a specific version as the storage version, bring it to the front of list of versions in the value of KUBE_API_VERSIONS.
+If you want to have more flexibility for configuring your cluster, you'll need to learn more about the [Salt stack](http://saltstack.com/).
 
-### Switching your config files to a new API version
+## Upgrading a cluster
 
-You can use the kube-version-change utility to convert config files between different API versions.
+GKE handles cluster upgrades for you, so you don't need to worry about. If you already a cluster running in some other environment and want to upgrade Kubernetes version running on it, you can use the `cluster/kube-push.sh` script. It works out of the box if your cluster runs on GCE and most of other providers.
```
Comment (thockin) [inline docs/cluster_management.md]: I think more details about how to use kube-push would be the most useful part of this doc.

File: docs/cluster_management.md [hunk 3]
```diff
@@ -2,45 +2,76 @@
...
 
+If your cluster is running on GCE and you have a repository with the version of Kubernetes you want to push to your cluster, you just need to do the following:
```
Comment (thockin) [inline docs/cluster_management.md]: What if I don't want to push from head, but want to push a new release binary?

File: docs/cluster_management.md [hunk 4]
```diff
@@ -2,45 +2,76 @@
...
 ```
+hack/dev-build-and-push.sh
+```
+This will build new version of Kubernetes, run basic unit and integration tests on it, and if all tests pass upgrade running Kubernetes to the version you just built.
+
+## Scaling the running cluster
+
+When the Kubelet is given masters address it'll self register in running Kubernetes cluster.
+
+If your cluster runs out short on resources you can easily add more machines to it. If you're using GCE or GKE it is done by resizing Instance Group managing your Nodes. It's done by modifying number of instances on `Compute > Compute Engine > Instance groups > your group > Edit group` page. Instance Group will take care of putting appropriate image on new machines and start them, while Kubelet will register its Node with API server to make it available for scheduling.
```
Comment (thockin) [inline docs/cluster_management.md]: You should link to cloud-console or something here - it's no clear what "page" you mean.

File: docs/cluster_management.md [hunk 5]
```diff
@@ -2,45 +2,76 @@
...
+
+In other environments you may need to configure the machine yourself and tell the Kubelet on which machine API server is running.
+
+## Cluster introspection
+
+To learn about the state of the cluster you can use a number of available tools. The simplest one is plain `kubectl` which can `describe node <node_name>`, giving you a summary of all information Kubernetes has about this particular Node. Other available tools include Elasticsearch, Graphana etc. Kubernetes is shipped with a number of so called cluster addons that provide additional functionality, such as monitoring. They run as ordinary pods and can be managed in exactly the same way. More detailed description is available [here](../cluster/addons).
```
Comment (thockin) [inline docs/cluster_management.md]: For things like kubectl commands, it's better to list a full example and be more specific about what they can see there.

File: docs/cluster_management.md [hunk 6]
```diff
@@ -2,45 +2,76 @@
...
+
+## Tearing down the cluster
+If you're running your cluster on GKE you can delete it from the web UI, or using `gcloud` CLI. Otherwise you may use `cluster/kube-down.sh` script which will try to turn down the cluster for you. On GCE `kube-down.sh` will delete all resources, but depending on the environment you're running in, you may need to clean up some things yourself, such as some network leftovers, persistent storage etc.
```
Comment (thockin) [inline docs/cluster_management.md]: can I use kube-push and kube-down on GKE?  For this doc to be really useful it needs to be precise and specific.  "may need to" is not great.  Explaining what they might need to do, and under what circumstances is better.

**Comment by thockin** [comment]
Thanks for getting on this.  Ideally a doc like this is full of cut-and-paste-able commandlines that users can use to do the real things they want to do.

**Comment by gmarek** [comment]
I didn't manage to look into it today, sorry about that. I may have some time this afternoon, but I may end up doing it tomorrow.

File: docs/cluster_management.md (hunk 1 above)
Comment (gmarek) [inline docs/cluster_management.md]: Nope, I didn't knew about it. Taking a look now.

File: docs/cluster_management.md (hunk 2 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done.

File: docs/cluster_management.md (hunk 3 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done.

File: docs/cluster_management.md (hunk 5 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done.

**Comment by gmarek** [comment]
PTAL and ignore k8s-bot...

**Comment by gmarek** [comment]
@thockin @bgrant0607

File: docs/cluster_management.md [hunk 7]
```diff
@@ -2,45 +2,152 @@
...
+
 ## Upgrading a cluster
 
-The `cluster/kube-push.sh` script will do a rudimentary update; it is a 1.0 roadmap item to have a robust live cluster update system.
+GKE handles cluster upgrades for you, so you don't need to worry about. If you already have a cluster running in some other environment and want to upgrade Kubernetes version running on it, you can use the `cluster/kube-push.sh` script. It works out of the box if your cluster runs on GCE and most of other providers. The upgrade is in-place, meaning that all user pods are kept alive throught this process.
```
Comment (derekwaynecarr) [inline docs/cluster_management.md]: GKE handles cluster upgrades for you, so you don't need to worry about it.

**Comment by bgrant0607** [comment]
cc @lavalamp @erictune

File: docs/cluster_management.md [hunk 8]
```diff
@@ -2,45 +2,152 @@
 
 This doc is in progress.
```
Comment (erictune) [inline docs/cluster_management.md]: remove "This doc is in progress"

File: docs/cluster_management.md [hunk 9]
```diff
@@ -2,45 +2,152 @@
...
 
+## Overview
+
+Kubernetes is an orchestration system for managing container clusters. This document provides basic instructions for setting up and managing the cluster and assumes familiarity with Kubernetes architecture and terminology.
```
Comment (erictune) [inline docs/cluster_management.md]: Overview not needed.  This doc should be linked-to by some other higher-level doc which gives enough context.

File: docs/cluster_management.md [hunk 10]
```diff
@@ -2,45 +2,152 @@
...
+
+## Prerequisites
+
+Kubernetes is written in Go and with some work it should be possible to run it anywhere where Docker can run (yes, this includes Raspberry Pi). Due to differences between Linux distributions and between cloud providers setting it up requires different amount of work on different environments. The best and simplest experience is provided by using Google Container Engine [GKE](https://cloud.google.com/container-engine/) or Google Compute Engine [GCE](https://cloud.google.com/compute/). Other environments are supported and maintained by the community, Vagrant or CoreOS on various IaaS providers being two examples.
```
Comment (erictune) [inline docs/cluster_management.md]: Not needed.  Covered by docs/getting-started-guide/README.md and docs/getting-started-guide/scratch.md

File: docs/cluster_management.md [hunk 11]
```diff
@@ -2,45 +2,152 @@
...
+
+## Creating and configuring a Cluster
+
+Assuming that you have machines for your clusters ready, you need to install Kubernetes on them. For specific instructions you can consult one of existing Getting Started guides for various environments:
+* GKE [GKE page](https://cloud.google.com/container-engine),
+* Vagrant [Vagrant setup](../docs/getting-started-guides/vagrant.md),
+* GCE [GCE setup](../docs/getting-started-guides/gce.md),
+* CoreOS on Azure [CoreOS on Azure setup](../docs/getting-started-guides/coreos/azure/README.md),
+* CoreOS on AWS [CoreOS on AWS setup](../docs/getting-started-guides/coreos.md),
+* CoreOS on VMWare [CoreOS on VMWare setup](../docs/getting-started-guides/coreos.md).
```
Comment (erictune) [inline docs/cluster_management.md]: Link directory to docs/getting-started-guide/README.md?

File: docs/cluster_management.md [hunk 12]
```diff
@@ -2,45 +2,152 @@
...
+
+For more comprehensive and detailed description of cluster setup and configuration process see [From Scratch](../docs/getting-started-guides/scratch.md) guide.
+
+## Difference between user-oriented and ops-oriented tools
+
+Currently Kubernetes has two kinds of tools: developer oriented yet useful for operations people `kubectl` or `cluster/kubectl.sh` that use user specified configuration/contexts. `kubectl` is a stable version of `cluster/kubectl.sh` distributed with the `gcloud` stack.
```
Comment (erictune) [inline docs/cluster_management.md]: we are not encouraging people to use cluster/kubectl.sh.  They should just install kubectl in their path.  Also, kubectl is an OSS thing, not a gcloud-specific thing.

File: docs/cluster_management.md [hunk 13]
```diff
@@ -2,45 +2,152 @@
...
+
+Cluster configuration/management tools stored in the `cluster` and `hack` folder, which are targeted to operations people, and only people managing clusters should use them. Those tools are mainly `bash` scripts which make managing single cluster easier, but they are currently not designed to allow easy management of multiple clusters from a single machine. If it's necessary it is possible to replicate existing e2e test cluster setup that consists mainly of `hack/ginkgo-e2e.sh` script and special test configs for specific cloud providers e.g. `cluster/gce/config-test.sh`. This means that in general scripts from `cluster` directory find a running cluster by themselves, and you don't need to explicitly specify the cluster you want to modify, as those tools assume that there's the cluster running.
```
Comment (erictune) [inline docs/cluster_management.md]: Actually, now everything in hack is for developers.  You can be an admin but not be a developer.  There will probably be many more admins than devs if we are doing it right.

Comment (erictune) [inline docs/cluster_management.md]: No need to mention `hack/ginkgo-e2e.sh` in an admin-oriented doc.

File: docs/cluster_management.md [hunk 14]
```diff
@@ -2,45 +2,152 @@
...
+
 ## Upgrading a cluster
 
-The `cluster/kube-push.sh` script will do a rudimentary update; it is a 1.0 roadmap item to have a robust live cluster update system.
+GKE handles cluster upgrades for you, so you don't need to worry about. If you already have a cluster running in some other environment and want to upgrade Kubernetes version running on it, you can use the `cluster/kube-push.sh` script. It works out of the box if your cluster runs on GCE and most of other providers. The upgrade is in-place, meaning that all user pods are kept alive throught this process.
 
-## Updgrading to a different API version
+`kube-push.sh` has a few functionalities. It can either be used to upgrade all Kubernetes components to the currently built binaries (default option), or to do limited upgrade of either the master machine or a single node. Instead of upgrading to the head, you can also specify the version which you want to roll out.
```
Comment (erictune) [inline docs/cluster_management.md]: Things to mention here:
 - example of command line to use to upgrade using kube-push.sh to upgrade to a specific released version.  Most people should be using released binaries.  e.g. setting KUBE_VERSION.
 - call out the fact that kube-push.sh can only be used for certain distros (some of the ones listed under `cluster/`, which isn't all of the distros.  And some under there may not support all the steps that kube-push needs).
- ability to upgrade master separately from nodes (see usage string in kube-push.sh)
- need to set KUBERNETES_PROVIDER .

Comment (erictune) [inline docs/cluster_management.md]: I'm starting to think that upgrade documentation belongs separately in each getting-started-guide.

Comment (erictune) [inline docs/cluster_management.md]: Okay, I see you have some examples below.  Some of my comments still apply

File: docs/cluster_management.md [hunk 15]
```diff
@@ -2,45 +2,152 @@
...
 
-There is a sequence of steps to upgrade to a new API version.
+If you made some changes to the release you're running, or downloaded the head repository, and want to upgrade all components, you can run following commands from the `kubernetes` directory:
+```
+make release
```
Comment (erictune) [inline docs/cluster_management.md]: Let's not talk about "make release" in this document.

File: docs/cluster_management.md (hunk 7 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done.

File: docs/cluster_management.md (hunk 8 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done.

File: docs/cluster_management.md (hunk 9 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done. Added a link to docs/admin (which links to this document) to main Kubernetes Readme.md

File: docs/cluster_management.md (hunk 10 above)
Comment (gmarek) [inline docs/cluster_management.md]: Removed.

File: docs/cluster_management.md (hunk 11 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done.

File: docs/cluster_management.md (hunk 12 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done.

File: docs/cluster_management.md (hunk 13 above)
Comment (gmarek) [inline docs/cluster_management.md]: Yeah - it's the only example I was able to find which shows how one can manage two clusters from one machine. Should I inline instructions here?

File: docs/cluster_management.md (hunk 14 above)
Comment (gmarek) [inline docs/cluster_management.md]: Done (I hope). As for putting upgrade documentation to separate docs - I think it would be a failure if we had to do it.

File: docs/cluster_management.md (hunk 15 above)
Comment (gmarek) [inline docs/cluster_management.md]: IIUC it's because you want to have an 'admin-not-dev' guide, right? I'll comply, but I'm not really sure I agree with very sharp distinction.

**Comment by gmarek** [comment]
PTAL

**Comment by gmarek** [comment]
@erictune PTAL

**Comment by bgrant0607** [comment]
@lavalamp @etune What should we do with this?

**Comment by bgrant0607** [comment]
cc @davidopp

**Comment by davidopp** [comment]
I can review this, but it is going to conflict with #11402.

**Comment by bgrant0607** [comment]
I suggested incorporating any useful parts of this into that PR.

**Comment by davidopp** [comment]
I actually think this one has a greater chance to get in in time than the other one, since @gmarek has a full work day before the doc copy.

File: README.md [hunk 16]
```diff
@@ -57,6 +57,10 @@ Kubernetes documentation is organized into several categories.
     - in [docs](docs/overview.md)
     - for people who want to run programs on kubernetes
     - describes current features of the system (with brief mentions of planned features)
+  - **Admin Documentation**
```
Comment (davidopp) [inline README.md]: Thanks for noticing this. To make sure this gets in before the deadline, I created a separate PR with this change (#11560)

File: docs/admin/README.md [hunk 17]
```diff
@@ -49,10 +49,10 @@ uses OpenVSwitch to set up networking between pods across
   Kubernetes nodes.
 
 If you are modifying an existing guide which uses Salt, this document explains [how Salt is used in the Kubernetes
-project.](salt.md).
+project](salt.md).
 
-## Upgrading a cluster
-[Upgrading a cluster](cluster-management.md).
+## Managing a cluster
```
Comment (davidopp) [inline docs/admin/README.md]: Thanks for noticing this. To make sure this gets in before the deadline, I created a separate PR with this change (#11560)

File: docs/admin/cluster-management.md [hunk 18]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
 <!-- END MUNGE: UNVERSIONED_WARNING -->
 # Cluster Management
 
-This doc is in progress.
+## Creating and configuring a Cluster
+
+Assuming that you have machines for your clusters ready, you need to install Kubernetes on them. For specific instructions you can consult one of existing [Getting Started guides](../../docs/getting-started-guides/README.md) for various environments:
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: I would remove everything in this subsection after this sentence. It will be annoying to try to keep this list up-to-date, and the From Scratch doc is already pointed to from docs/getting-started-guides/README.md

File: docs/admin/cluster-management.md [hunk 19]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
+* GKE [GKE page](https://cloud.google.com/container-engine),
+* Vagrant [Vagrant setup](../../docs/getting-started-guides/vagrant.md),
+* GCE [GCE setup](../../docs/getting-started-guides/gce.md),
+* CoreOS on Azure [CoreOS on Azure setup](../../docs/getting-started-guides/coreos/azure/README.md),
+* CoreOS on AWS [CoreOS on AWS setup](../../docs/getting-started-guides/coreos.md),
+* CoreOS on VMWare [CoreOS on VMWare setup](../../docs/getting-started-guides/coreos.md).
+
+For more comprehensive and detailed description of cluster setup and configuration process see [From Scratch](../../docs/getting-started-guides/scratch.md) guide.
+
+## Tools for interacting with Kubernetes
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: I think you should remove this section. The first paragraph is not needed because we describe kuebctl in the user guide, and the getting started guides explain when/how to use the relevant commands from cluster/. The stuff in the second and third paragraph is good but should be merged into cluster/README.md, not put in this doc.

File: docs/admin/cluster-management.md [hunk 20]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
+Note the `cluster/kube-env.sh` script that defines the `KUBERNETES_PROVIDER` variable which is used by all other scripts. Default value is `gce` so if you want to manage non-GCE cluster you need to assign an identifier of your provider (e.g. `gke`, `aws`, `azure`, `vagrant`, etc.) to `KUBERNETES_PROVIDER` variable e.g. by putting the line
+```bash
+export KUBERNETES_PROVIDER=vagrant
+```
+to your `.bashrc` or `.profile` file.
 
 ## Upgrading a cluster
 
-The `cluster/kube-push.sh` script will do a rudimentary update; it is a 1.0 roadmap item to have a robust live cluster update system.
+GKE handles cluster upgrades for you, so you don't need to worry about it. If you already have a cluster running in some other environment and want to upgrade Kubernetes version running on it, you can use the `cluster/kube-push.sh` script. It works out of the box if your cluster runs on GCE and some other providers listed in `cluster` folder. Note that `kube-push.sh` requires that cloud providers implement necessary interface, thus not all of them are currently supported.
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: I would remove this from your PR for now, as @brendandburns has another PR in flight with conflicting changes to this  section (#11402). Once his is in, you could add more detail as you see fit (but given the timing, I guess that wouldn't make it into the 1.0 documentation). Also, why did you remove all of the information on upgrading API version?

File: docs/admin/cluster-management.md [hunk 21]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
+```
+cluster/kube-push.sh -n my_node v0.20.2
+``` 
 
-1. Turn on the new api version
-2. Upgrade the cluster's storage to use the new version.
-3. Upgrade all config files. Identify users of the old api version endpoints.
-4. Update existing objects in the storage to new version by running cluster/update-storage-objects.sh
-3. Turn off the old version.
+## Scaling the running cluster
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: This is good information, but I think it belongs in docs/admin/node.md near the section "Self-Registration of nodes"

File: docs/admin/cluster-management.md [hunk 22]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
-### Turn on or off an API version for your cluster
+In other environments you may need to configure the machine yourself and tell the Kubelet on which machine API server is running.
+TODO: flesh it out with some concrete example e.g. on Vagrant.
 
-Specific API versions can be turned on or off by passing --runtime-config=api/<version> flag while bringing up the server. For example: to turn off v1 API, pass --runtime-config=api/v1=false.
-runtime-config also supports 2 special keys: api/all and api/legacy to control all and legacy APIs respectively. For example, for turning off all api versions except v1, pass --runtime-config=api/all=false,api/v1=true.
+## Cluster introspection
 
-### Switching your cluster's storage API version
+The basic way of checking the state of the cluster is to use `kubectl` (or `cluster/kubectl.sh`) tool. To use it you need to configure `kubectl` first. The process of doing it is described in this [doc](../../docs/sharing-clusters.md).
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: Drop this paragraph. I was told we're not supposed to refer to cluster/kubectl.sh, just kubectl. Also, the multi-cluster scenario is already described in docs/user-guide/prereqs.md

File: docs/admin/cluster-management.md [hunk 23]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
 
-KUBE_API_VERSIONS env var controls the API versions that are supported in the cluster. The first version in the list is used as the cluster's storage version. Hence, to set a specific version as the storage version, bring it to the front of list of versions in the value of KUBE_API_VERSIONS.
+You can use `kubectl` to get addresses of all running cluster addons and kubernetes master. E.g.
+```
+cluster/kubectl.sh cluster-info
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: cluster/kubectl.sh -> kubectl.sh

File: docs/admin/cluster-management.md (hunk 20 above)
Comment (davidopp) [inline docs/admin/cluster-management.md]: To clarify, when I say "I would remove this from your PR for now," I mean "I would remove all the changes you made to the Upgrading a cluster section" (and the deletions of the upgrading API version section)

File: docs/admin/cluster-management.md [hunk 24]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
+```
+Instance Group will take care of putting appropriate image on new machines and start them, while Kubelet will register its Node with API server to make it available for scheduling. If you scale the instance group down, system will randomly choose Nodes to kill, so if you want to keep everything running all the time you probably shouldn't downsize the cluster twice.
 
-### Turn on or off an API version for your cluster
+In other environments you may need to configure the machine yourself and tell the Kubelet on which machine API server is running.
+TODO: flesh it out with some concrete example e.g. on Vagrant.
 
-Specific API versions can be turned on or off by passing --runtime-config=api/<version> flag while bringing up the server. For example: to turn off v1 API, pass --runtime-config=api/v1=false.
-runtime-config also supports 2 special keys: api/all and api/legacy to control all and legacy APIs respectively. For example, for turning off all api versions except v1, pass --runtime-config=api/all=false,api/v1=true.
+## Cluster introspection
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: Actually, I would drop this whole section.
docs/user-guide/introspection-and-debugging.md
already talks about 'describe node' and
docs/user-guide/accessing-the-cluster.md
talks about 'cluster-info'
The one thing you might want to do is to add the info about 'get events' to
docs/user-guide/introspection-and-debugging.md

File: docs/admin/cluster-management.md [hunk 25]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
-$ hack/build-go.sh cmd/kube-version-change
-$ _output/local/go/bin/kube-version-change -i myPod.v1beta3.yaml -o myPod.v1.yaml
+but you have to remember that events are namespaced. This means that if you're interested in events for some namespaced object (e.g. what happened with pods in namespace `my-namespace`) you need to explicitly provide a namespace to this command:
 ```
+cluster/kubectl.sh get events --namespace=my-namespace
+```
+
+Other available introspection/monitoring tools include Elasticsearch, Graphana etc. They are shipped with Kubernetes to provide additional functionality, such as monitoring resource usage, or allowing searching logs. They run as ordinary pods and can be managed in exactly the same way. More detailed description of cluster addons is available [here](../../cluster/addons/).
+
+## Tearing down the cluster
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: Remove this section. This information is already covered in docs/getting-started-guides/
(except I guess GKE).

File: docs/admin/cluster-management.md [hunk 26]
```diff
@@ -14,47 +14,141 @@ certainly want the docs that go with that version.</h1>
...
+If you're running your cluster on GKE you can delete it from the web UI, or using `gcloud` CLI. 
+
+Otherwise you tear down the cluster by using `cluster/kube-down.sh` script. On GCE `kube-down.sh` will delete all resources, but depending on the environment you're running in, you may need to clean up some things yourself, such as some network leftovers, persistent storage etc.
 
 ### Maintenance on a Node
 
+TODO: this is out of date, need a rewrite as well.
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: I would drop the changes to this section (including the TODO) except for:
- "assuming that not all nodes will go down in the same time" is fine to keep, but change it to "at the same time"
- capitalizing Additionally

pod-eviction-timeout is mentioned in node.md (or it should be mentioned there), it is too much detail to put here.

**Comment by davidopp** [comment]
Hi Marek. These were good changes, but most of them are now obsolete based on other doc changes that were made in the two weeks since you created this PR. I have written detailed comments in the PR, but the short summary is that I would get rid of everything in this PR except
* The "Creating and configuring a Cluster" heading and the first sentence of that section
* two changes to "Maintenance on a node" section (see my comment on that section)

In a couple of cases you wrote stuff that is not obsolete, but just belongs in other docs. See my detailed comments for more information.

File: docs/admin/cluster-management.md (hunk 19 above)
Comment (gmarek) [inline docs/admin/cluster-management.md]: Done.

File: docs/admin/cluster-management.md (hunk 18 above)
Comment (gmarek) [inline docs/admin/cluster-management.md]: Done.

File: docs/admin/cluster-management.md (hunk 20 above)
Comment (gmarek) [inline docs/admin/cluster-management.md]: Done.

File: docs/admin/cluster-management.md (hunk 21 above)
Comment (gmarek) [inline docs/admin/cluster-management.md]: I don't agree, as node.md holds only design information (for now), while this thing is about changing things (with example). I'll move it either way to make things go smooth.

**Comment by gmarek** [comment]
When I was writing this doc I was trying to follow what @RichieEscarez specified in #10125. It seems to me, that there was by far not enough communication about what's the goal/vision of this doc and I lost quite some time because of that.

File: docs/admin/cluster-management.md [hunk 27]
```diff
@@ -33,7 +33,9 @@ Documentation for other releases can be found at
 
 # Cluster Management
 
-This doc is in progress.
+## Creating and configuring a Cluster
+
+Assuming that you have machines for your clusters ready, you need to install Kubernetes on them. For specific instructions you can consult one of existing [Getting Started guides](../../docs/getting-started-guides/README.md) for various environments.
```
Comment (davidopp) [inline docs/admin/cluster-management.md]: Please change to just "To install Kubernetes on a set of machines, consult one of the existing [Getting Started guides](../../docs/getting-started-guides/README.md) depending on your environment."

File: docs/admin/node.md [hunk 28]
```diff
@@ -191,6 +191,14 @@ For self-registration, the kubelet is started with the following options:
 Currently, any kubelet is authorized to create/modify any node resource, but in practice it only creates/modifies
 its own.  (In the future, we plan to limit authorization to only allow a kubelet to modify its own Node resource.)
 
+If your cluster runs out short on resources you can easily add more machines to it if your cluster is running in Node self-registartion mode. If you're using GCE or GKE it's done by resizing Instance Group managing your Nodes. It can be accomplished by modifying number of instances on `Compute > Compute Engine > Instance groups > your group > Edit group` [Google Cloud Console page](https://console.developers.google.com) or using gcloud CLI:
```
Comment (davidopp) [inline docs/admin/node.md]: If your cluster runs out short -> If your cluster runs short
self-registartion -> self-registration

File: docs/admin/node.md [hunk 29]
```diff
@@ -191,6 +191,14 @@ For self-registration, the kubelet is started with the following options:
...
+```
+gcloud preview managed-instance-groups --zone compute-zone resize my-cluster-minon-group --new-size 42
+```
+Instance Group will take care of putting appropriate image on new machines and start them, while Kubelet will register its Node with API server to make it available for scheduling. If you scale the instance group down, system will randomly choose Nodes to kill, so if you want to keep everything running all the time you probably shouldn't downsize the cluster twice.
```
Comment (davidopp) [inline docs/admin/node.md]: remove "so if you want to keep everything running all the time you probably shouldn't downsize the cluster twice."

File: docs/user-guide/introspection-and-debugging.md [hunk 30]
```diff
@@ -181,6 +181,15 @@ Here you can see the event generated by the scheduler saying that the Pod failed
 
 To correct this situation, you can use `kubectl scale` to update your Replication Controller to specify four or fewer replicas. (Or you could just leave the one Pod pending, which is harmless.)
 
+Events such as ones you saw at the end of `kubectl describe pod` are persisted in the etcd and provide high level information on what was happening in the cluster. To list all events you can use
```
Comment (davidopp) [inline docs/user-guide/introspection-and-debugging.md]: the etcd -> etcd
high level -> high-level
was happening -> is happening

File: docs/user-guide/introspection-and-debugging.md [hunk 31]
```diff
@@ -181,6 +181,15 @@ Here you can see the event generated by the scheduler saying that the Pod failed
...
+```
+kubectl get events
+```
+but you have to remember that events are namespaced. This means that if you're interested in events for some namespaced object (e.g. what happened with Pods in namespace `my-namespace`) you need to explicitly provide a namespace to this command:
```
Comment (davidopp) [inline docs/user-guide/introspection-and-debugging.md]: this command -> the command

File: docs/admin/cluster-management.md (hunk 27 above)
Comment (gmarek) [inline docs/admin/cluster-management.md]: Hmm.. I thought I removed this all together...

File: docs/admin/node.md (hunk 28 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/admin/node.md (hunk 29 above)
Comment (gmarek) [inline docs/admin/node.md]: Done.

File: docs/user-guide/introspection-and-debugging.md (hunk 30 above)
Comment (gmarek) [inline docs/user-guide/introspection-and-debugging.md]: Done.

File: docs/user-guide/introspection-and-debugging.md (hunk 31 above)
Comment (gmarek) [inline docs/user-guide/introspection-and-debugging.md]: Done.

File: docs/user-guide/introspection-and-debugging.md [hunk 32]
```diff
@@ -181,6 +181,15 @@ Here you can see the event generated by the scheduler saying that the Pod failed
...
 To correct this situation, you can use `kubectl scale` to update your Replication Controller to specify four or fewer replicas. (Or you could just leave the one Pod pending, which is harmless.)
 
+Events such as ones you saw at the end of `kubectl describe pod` are persisted in etcd and provide high-level information on what is happening in the cluster. To list all events you can use
+```
+kubectl get events
+```
+but you have to remember that events are namespaced. This means that if you're interested in events for some namespaced object (e.g. what happened with Pods in namespace `my-namespace`) you need to explicitly provide a namespace to the command:
+```
+kubectl get events --namespace=my-namespace
+```
```
Comment (davidopp) [inline docs/user-guide/introspection-and-debugging.md]: Can you add

To see events from all namespaces, you can use the `--all-namespaces` argument

File: docs/user-guide/introspection-and-debugging.md [hunk 33]
```diff
@@ -181,6 +181,15 @@ Here you can see the event generated by the scheduler saying that the Pod failed
 
 To correct this situation, you can use `kubectl scale` to update your Replication Controller to specify four or fewer replicas. (Or you could just leave the one Pod pending, which is harmless.)
 
+Events such as ones you saw at the end of `kubectl describe pod` are persisted in etcd and provide high-level information on what is happening in the cluster. To list all events you can use
```
Comment (davidopp) [inline docs/user-guide/introspection-and-debugging.md]: ones -> the ones

File: docs/user-guide/introspection-and-debugging.md (hunk 32 above)
Comment (gmarek) [inline docs/user-guide/introspection-and-debugging.md]: Done.

File: docs/user-guide/introspection-and-debugging.md (hunk 33 above)
Comment (gmarek) [inline docs/user-guide/introspection-and-debugging.md]: Done.

**Comment by davidopp** [comment]
LGTM

Sorry for the miscommunication about the other doc changes that were going on.
